### PR TITLE
websocket restructure and multiple connections changes

### PR DESCRIPTION
**Pull request by Dileepnidi**
websocket restructure and multiple connections changes

### CODE REVIEW

File: ReactSkia/sdk/NopollWebsocket.cpp [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+    } else {
+      globalMtx.lock();
+      currentWorkerThreadCount--;
+      globalMtx.unlock();
+      requestQueueMtx.unlock();
+      break;
+    }
+    requestQueueMtx.unlock();
+    if(!nopollRequest)
+      return;
```
Comment (pprabhu-ng) [inline ReactSkia/sdk/NopollWebsocket.cpp]: this should be continue instead of return ?

File: ReactSkia/sdk/NopollWebsocket.cpp [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+namespace react {
+
+std::queue<NopollRequest*> RequestQueue;
+int currentWorkerThreadCount = 0;
+std::mutex globalMtx;
+std::mutex requestQueueMtx;
+sem_t startDownloadingThread;  /*semaphore to signal downloading thread to start, once there is something in the list*/
+
+NopollWebsocket* NopollWebsocket::sharedNopollWebsocket_{nullptr};
+std::mutex NopollWebsocket::mutex_;
```
Comment (pprabhu-ng) [inline ReactSkia/sdk/NopollWebsocket.cpp]: variable name can be requestLock ?

File: ReactSkia/sdk/NopollWebsocket.cpp [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+#include "NopollWebsocket.h"
+
+using namespace std;
+namespace facebook {
+namespace react {
+
+std::queue<NopollRequest*> RequestQueue;
+int currentWorkerThreadCount = 0;
+std::mutex globalMtx;
+std::mutex requestQueueMtx;
```
Comment (pprabhu-ng) [inline ReactSkia/sdk/NopollWebsocket.cpp]: requestQueueMtx  => requestQueueLock ?

File: ReactSkia/sdk/NopollWebsocket.cpp [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+#include "ReactSkia/utils/RnsLog.h"
+#include "NopollWebsocket.h"
+
+using namespace std;
+namespace facebook {
+namespace react {
+
+std::queue<NopollRequest*> RequestQueue;
+int currentWorkerThreadCount = 0;
+std::mutex globalMtx;
```
Comment (pprabhu-ng) [inline ReactSkia/sdk/NopollWebsocket.cpp]: globalMtx => threadCountLock ?

File: ReactSkia/sdk/NopollWebsocket.cpp (hunk 1 above)
Comment (Dileepnidi) [reply ReactSkia/sdk/NopollWebsocket.cpp]: changed

File: ReactSkia/sdk/NopollWebsocket.cpp (hunk 2 above)
Comment (Dileepnidi) [reply ReactSkia/sdk/NopollWebsocket.cpp]: changed to suggested name

File: ReactSkia/sdk/NopollWebsocket.cpp (hunk 3 above)
Comment (Dileepnidi) [reply ReactSkia/sdk/NopollWebsocket.cpp]: changed to suggested name

File: ReactSkia/sdk/NopollWebsocket.cpp (hunk 4 above)
Comment (Dileepnidi) [reply ReactSkia/sdk/NopollWebsocket.cpp]: changed to suggested name

File: ReactSkia/sdk/NopollWebsocket.cpp [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+
+#include "ReactSkia/utils/RnsLog.h"
+#include "NopollWebsocket.h"
+
+using namespace std;
+namespace facebook {
+namespace react {
+
+std::queue<NopollRequest*> RequestQueue;
+int currentWorkerThreadCount = 0;
```
Comment (munezbn) [inline ReactSkia/sdk/NopollWebsocket.cpp]: Can we make this as part of singleton class object (static member may be) ? Same for other global variables.

File: ReactSkia/sdk/NopollWebsocket.cpp [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+    } else {
+      threadCountLock.lock();
+      currentWorkerThreadCount--;
+      threadCountLock.unlock();
+      requestQueueLock.unlock();
+      break;
+    }
+    requestQueueLock.unlock();
+    if(!nopollRequest)
+      continue;
```
Comment (munezbn) [inline ReactSkia/sdk/NopollWebsocket.cpp]: I think better to move this check after we call front() on queue

File: ReactSkia/sdk/NopollWebsocket.cpp [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+}
+
+void NopollWebsocket::multiConn(){
+  NopollRequest* nopollRequest = nullptr;
+  while(true) { // forever
+    requestQueueLock_.lock();
+    if(RequestQueue_.size() !=0) {
+      nopollRequest =RequestQueue_.front();
+      if(!nopollRequest)
+        continue;
```
Comment (pprabhu-ng) [inline ReactSkia/sdk/NopollWebsocket.cpp:58]: dont we need to unlock before calling continue, else we wont be able to acquire the lock in the next loop

File: ReactSkia/sdk/NopollWebsocket.cpp (hunk 5 above)
Comment (Dileepnidi) [reply ReactSkia/sdk/NopollWebsocket.cpp]: moved all global variables to private

File: ReactSkia/sdk/NopollWebsocket.cpp (hunk 6 above)
Comment (Dileepnidi) [reply ReactSkia/sdk/NopollWebsocket.cpp]: moved.

File: ReactSkia/sdk/NopollWebsocket.cpp (hunk 7 above)
Comment (Dileepnidi) [reply ReactSkia/sdk/NopollWebsocket.cpp:58]: updated

**Review by pprabhu-ng — APPROVED** [review]

**Review by munezbn — APPROVED** [review]

**Review by chatrapathi06 — APPROVED** [review]

**Review by munezbn — APPROVED** [review]

File: ReactSkia/modules/platform/nopoll/RSkWebSocketModule.cpp [hunk 8]
```diff
@@ -13,227 +12,142 @@
...
-    }
-	        
-  parameters["id"] = socketID;
-  parameters["message"] = "sending data is failed";
-  sendEventWithName(events_[3], folly::dynamic(parameters));
-  RNS_LOG_ERROR("sending data is failed");
-  return jsi::Value(WEBSOCKET_RETURN_FAILURE);
-
+  connectionListLock_.lock();
+	WebsocketRequest*  websocketRequest =  connectionList_[socketID];
```
Comment (pprabhu-ng) [inline ReactSkia/modules/platform/nopoll/RSkWebSocketModule.cpp]: indentation to be corrected here

File: ReactSkia/modules/platform/nopoll/RSkWebSocketModule.cpp [hunk 9]
```diff
@@ -13,227 +12,142 @@
...
-  parameters = folly::dynamic::object();
-  parameters["id"] = socketID;
-  parameters["message"] = "sending binary data is failed";
-  sendEventWithName(events_[3], folly::dynamic(parameters));
-  RNS_LOG_ERROR("sending binary data is failed");
-  return jsi::Value(WEBSOCKET_RETURN_FAILURE);
-
-
+  connectionListLock_.lock();
+	WebsocketRequest*  websocketRequest =  connectionList_[socketID];
```
Comment (pprabhu-ng) [inline ReactSkia/modules/platform/nopoll/RSkWebSocketModule.cpp]: indentation to be corrected here

File: ReactSkia/modules/platform/nopoll/RSkWebSocketModule.cpp (hunk 8 above)
Comment (Dileepnidi) [reply ReactSkia/modules/platform/nopoll/RSkWebSocketModule.cpp]: updated

File: ReactSkia/modules/platform/nopoll/RSkWebSocketModule.cpp (hunk 9 above)
Comment (Dileepnidi) [reply ReactSkia/modules/platform/nopoll/RSkWebSocketModule.cpp]: updated

**Review by pprabhu-ng — APPROVED** [review]

**Pull request merged**